# This patch is based on SphinxSearch REL1_27.tar.gz
# We remove a regex caret to enable substring searches of the title.

# Special for 1.35.14 upgrade: we also have to patch the DB instance logic, as 1.35 changed
# behavior half way through it's life (around 1.35.6), but the REL1_35 release of SphinxSearch
# wasn't updated.

--- SphinxMWSearch.php.orig  2025-04-30 14:53:36.624343273 -0700
+++ SphinxMWSearch.php        2025-04-30 14:54:45.491572842 -0700
@@ -68,12 +68,13 @@
         * @return bool
         */
        public static function prefixSearch( $namespaces, $term, $limit, &$results, $offset = 0 ) {
-               $term = '^' . $term;
+               // Displayed for ropewiki substring title matching
+               //$term = '^' . $term;
                return self::infixSearch( $namespaces, $term, $limit, $results, $offset );
        }

        public static function infixSearch( $namespaces, $term, $limit, &$results, $offset = 0 ) {
-               $search_engine = new SphinxMWSearch( wfGetDB( DB_REPLICA ) );
+               $search_engine = new SphinxMWSearch( MediaWiki\MediaWikiServices::getInstance()->getDBLoadBalancer() );
                $search_engine->namespaces = $namespaces;
                $search_engine->setLimitOffset( $limit, $offset );
                $result_set = $search_engine->searchText( '@page_title: ' . $term . '*' );

--- SphinxMWSearchResult.php.orig	2025-06-11 04:04:00.089677897 +0000
+++ SphinxMWSearchResult.php	2025-06-24 01:20:29.530809534 +0000
@@ -25,7 +25,20 @@
 	 *
 	 * @return string highlighted text snippet
 	 */
-	public function getTextSnippet( $terms ) {
+	public function getTextSnippet($terms = []) {
+		global $wgRequest;
+
+		// If $terms wasn't provided, try and figure it out from the url parameters
+		if ( empty( $terms ) ) {
+        	  $query = trim( $wgRequest->getVal( 'search', '' ) );
+	          if ( $query === '' ) {
+	            $query = trim( $wgRequest->getVal( 'q', '' ) );
+	          }
+	          if ( $query !== '' ) {
+	            $terms = preg_split( '/\s+/', $query );
+	          }
+		}
+
 		global $wgAdvancedSearchHighlighting, $wgSphinxSearchMWHighlighter, $wgSphinxSearch_index;
 
 		$this->initText();
@@ -48,10 +61,17 @@
 			"around" => $contextchars,
 		];
 
+		$query = trim( implode( ' ', $terms ) );
+		// It crashes searchd if the query is empty (null pointer), so fallback to ' '.
+		if ( $query === '' ) {
+		  $query = ' ';
+		}
+
+		wfDebugLog( 'SphinxSearch', 'BuildExcerpts query: ' . var_export( $terms, true ) );
 		$excerpts = $this->sphinx_client->BuildExcerpts(
 			[ $this->mText ],
 			$wgSphinxSearch_index,
-			implode( ' ', $terms ),
+			$query,
 			$excerpts_opt
 		);
 
